clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3556,6 +3590,37 @@ void Sema::AddSYCLIntelMaxWorkGroupSizeAttr(Decl *D,
   if (!XDim || !YDim || !ZDim)
     return;
 
+  // If the [[intel::max_work_group_size(X, Y, Z)]] attribute is specified on
+  // a declaration along with [[sycl::reqd_work_group_size(X1, Y1, Z1)]]
+  // attribute, check to see if values of reqd_work_group_size arguments are
+  // equal or less than values of max_work_group_size attribute arguments.
+  //
+  // The arguments to reqd_work_group_size are ordered based on which index
+  // increments the fastest. In OpenCL, the first argument is the index that
+  // increments the fastest, and in SYCL, the last argument is the index that
+  // increments the fastest.
+  //
+  // The arguments to max_work_group_size are ordered based on which index
+  // increments the fastest. In SYCL, the last argument is the index that
+  // increments the fastest.
+  if (const auto *DeclAttr = D->getAttr<ReqdWorkGroupSizeAttr>()) {
+    bool checkFirstArgument = getLangOpts().OpenCL
+              ? checkWorkGroupSizeAttrValues(DeclAttr->getXDim(), XDim)
+              : checkWorkGroupSizeAttrValues(DeclAttr->getZDim(), XDim);
+    bool checkSecondArgument =
+              checkWorkGroupSizeAttrValues(DeclAttr->getYDim(), YDim);
+    bool checkThirdArgument = getLangOpts().OpenCL
+              ? checkWorkGroupSizeAttrValues(DeclAttr->getZDim(), ZDim)
+              : checkWorkGroupSizeAttrValues(DeclAttr->getXDim(), ZDim);
+
+    if (checkFirstArgument && checkSecondArgument && checkThirdArgument) {
```

Comment:
This would give you wrong results.

checkWorkGroupSizeAttrValues returns true if rwgs is greater than mwgs, i.e. a violation.
So you need:
```suggestion
    if (checkFirstArgument || checkSecondArgument || checkThirdArgument) {
```
i.e. you want to issue the diagnostic if any of the values violate the rule.